src/stat.h

```diff
@@ -11,15 +11,61 @@
 #ifndef SQUID_STAT_H_
 #define SQUID_STAT_H_
 
+// TODO: Rephrase to support negative tops (and bottoms?) as in byte hit ratios.
+/// Maintains totals for events that are divided into two mutually exclusive
+/// categories (e.g., cache hits/misses or busy/idle time). One of the two
+/// categories is treated as "primary" for the purposes of computing event
+/// ratios (e.g., hit ratio or busy time percentage). Each event is associated
+/// with an integer "value" or "weight".
+class EventRatio
+{
+public:
+    using Value = double;
+
+    /// no events have been observed (e.g., hit ratio after zero requests)
+    EventRatio() = default;
+
+    /// \param primary is the cumulative weight of events in primary category
+    /// \param total is the cumulative weight of events in all categories
+    EventRatio(const Value primary, const Value total): primary_(primary), total_(total) {}
+
+    // TODO: Check whether this can be removed.
+    /// whether any events have been observed, allowing this ratio to be
+    /// accurately represented by a single non-negative number
+    explicit operator bool() const { return bool(total_); }
+
+    /// \returns 100*primary/total for non-zero totals
+    /// \returns noEventsValue for zero totals
+    double toPercentOr(double noEventsValue) const { return total_ ? (100.0*primary_)/total_ : noEventsValue; }
+
+// XXX private:
+    Value primary_ = 0;
+    Value total_ = 0;
+};
+
+inline EventRatio &
+operator +=(EventRatio &r1, const EventRatio &r2)
+{
+    // "Tops and bottoms" addition. The usual fractions addition does not work
+    // for event ratios. For example, if both workers received only hits, then
+    // an SMP Squid instance document hit ratio is 100% rather than 200% that
+    // adding n/n and m/m fractions would have yielded. Similarly, dividing
+    // the usual fractions sum by the number of fractions yields wrong ratios:
+    // (1000/1000 + 1/10)/2 = 55% compared to ~99% actual instance hit ratio.
```

Comment:
I'd suggest to refer to the existing [weighted mean](https://en.wikipedia.org/wiki/Weighted_arithmetic_mean) concept instead of this detailed explanation.